v1/lists.js

```diff
@@ -0,0 +1,190 @@
+     */
+    unflattenContinuation(continuation) {
+        const sanitizedContinuation = {};
+        if (typeof continuation === 'string') {
+            try {
+                continuation = JSON.parse(continuation);
+            } catch (e) {
+                this.options.log('error/unflatten', {
+                    msg: e.stack,
+                    json: continuation,
+                });
+                throw e;
+            }
+            // Make sure nothing malicious can be done by splicing the continuation data
+            // into the API parameters.
+            const allowedKeys = ['continue', 'rlcontinue', 'rlecontinue'];
+            for (const key of allowedKeys) {
+                if (typeof continuation[key] !== 'object') {
+                    sanitizedContinuation[key] = continuation[key];
+                }
+            }
+        }
+        return sanitizedContinuation;
+    }
+
+    /**
+     * Convert an array of values into the format expected by the MediaWiki API.
+     * @param {!Array} list A list containing strings and numbers.
+     * @return {!String}
+     */
+    flattenMultivalue(list) {
+        return list.join('|');
+    }
+
+    /**
+     * Handle the /list/{id}/entries endpoint (get entries of a list).
+     * @param {!HyperSwitch} hyper
+     * @param {!Object} req The request object as provided by HyperSwitch.
+     * @return {!Promise<Object>} A response promise.
+     */
+    getListEntries(hyper, req) {
+        return hyper.post({
+            uri: new URI([req.params.domain, 'sys', 'action', 'rawquery']),
+            body: {
+                action: 'query',
+                list: 'readinglistentries',
+                rlelists: req.params.id,
+                rlelimit: 'max',
+                continue: this.unflattenContinuation(req.query.next).continue,
+                rlecontinue: this.unflattenContinuation(req.query.next).rlecontinue,
+            },
+        })
+        .then((res) => {
+            const entries = res.body.query.readinglistentries;
+            const next = this.flattenContinuation(res.body.continue);
+
+            return this.hydrateSummaries(entries)
+            .then((entries) => ({
+                status: 200,
+                headers: {
+                    'content-type': 'application/json; charset=utf-8;'
+                        + 'profile="https://www.mediawiki.org/wiki/Specs/Lists/0.1"',
+                    'cache-control': 'no-cache',
+                },
+                body: {
+                    entries,
+                    next,
+                }
+            }));
+        });
+    }
+
+    /**
+     * Add data from the summary endpoint to an array of list entries.
+     * @param {!Array} entries An array of objects with two properties:
+     *   - project: the domain of the wiki
+     *   - title: the title of the page
+     * @return {!Promise<Array>} The objects, enriched with summaries.
+     */
+    hydrateSummaries(entries) {
```

Comment:
For this you can use `hydrateResponse()` and `fetchSummary()` from `lib/mwUtil.js`. For a sample usage c.f. [how the feed end point does it](https://github.com/wikimedia/restbase/blob/564e23da394ae480ec03e286f9ccf3fcbacf37f9/v1/feed.js#L58).

Reply:
`fetchSummary()` uses HyperSwitch so AIUI it assumes a local endpoint; I need to use the summary endpoint belonging to `entry.project` here. `hydrateResponse()` seems to be intended for responses which already include the URL to be hydrated; if I add the `$merge` fields manually before calling it, I don't think I can simplify the code much. Also it doesn't have concurrency limits, and in theory we could be processing a thosand entries here (although the actual limits will be much lower until [T177462](https://phabricator.wikimedia.org/T177462) gets resolved).

I think the only extra feature `hydrateResponse()` has is deduplication, and reading lists cannot contain the same page twice.

Reply:
Actually concurrency limits seem like a good idea in general, I will add it l8r.

`fetchSummary` doesn't assume local endpoint, it will fetch whatever URI you provide it.